packages/suite/src/actions/suite/metadataActions.ts

```diff
@@ -710,136 +733,293 @@ export const addAccountMetadata =
         }
 
-        dispatch({
-            type: METADATA.SET_DATA,
-            payload: {
+        dispatch(
+            setMetadata({
+                fileName,
                 provider,
+                data: nextMetadata,
+            }),
+        );
+
+        // we might intentionally skip saving metadata content to persistent storage.
+        if (!save) return true;
+
+        await dispatch(
+            encryptAndSaveMetadata({
                 data: {
-                    [fileName]: nextMetadata,
+                    accountLabel: nextMetadata.accountLabel,
+                    outputLabels: nextMetadata.outputLabels,
+                    addressLabels: nextMetadata.addressLabels,
                 },
+                aesKey,
+                fileName,
+                provider,
+            }),
+        );
+        return true;
+    };
+
+/**
+ * Generate device master-key
+ * */
+export const setDeviceMetadataKey =
+    (encryptionVersion = METADATA.ENCRYPTION_VERSION) =>
+    async (dispatch: Dispatch, getState: GetState) => {
+        if (!getState().metadata.enabled) return;
+        const { device } = getState().suite;
+        if (!device || !device.state || !device.connected) return;
+
+        // master key already exists
+        // if (device.metadata.status === 'enabled') return;
+
+        const result = await TrezorConnect.cipherKeyValue({
+            device: {
+                path: device.path,
+                state: device.state,
+                instance: device.instance,
             },
+            useEmptyPassphrase: device.useEmptyPassphrase,
+            ...METADATA.ENCRYPTION_VERSION_CONFIGS[encryptionVersion],
         });
 
-        // we might intentionally skip saving metadata content to persistent storage.
-        if (!save) return true;
+        if (result.success) {
+            if (!getState().metadata.enabled) {
+                dispatch({
+                    type: METADATA.ENABLE,
+                });
+            }
 
-        const providerInstance = await dispatch(
-            getProviderInstance({ clientId: provider.clientId }),
-        );
+            const [stateAddress] = device.state.split('@'); // address@device_id:instance
+            const metaKey = metadataUtils.deriveMetadataKey(result.payload.value, stateAddress);
+            const fileName = `${metadataUtils.deriveFilenameForLabeling(
+                metaKey,
+                encryptionVersion,
+            )}`;
+            const aesKey = metadataUtils.deriveAesKey(metaKey);
 
-        if (!providerInstance) {
-            // provider should always be set here (see init)
-            return false;
+            dispatch({
+                type: METADATA.SET_DEVICE_METADATA,
+                payload: {
+                    deviceState: device.state,
+                    metadata: {
+                        ...device.metadata,
+                        status: 'enabled',
+                        [encryptionVersion]: {
```

Comment:
> if you fail to call cipherKeyValue for V1, the V2 keys are also erased.

this is definitely not what I wanted. 

>  I'd consider renaming

sure feel free to do it

Reply:
So the outcome of would not only depend on `result.success` but also on the presence of some keys? If you call that action the second time and you already have v2 keys and you cancel when v1 keys are obtained, the state should not be cleared and set to `cancelled`? Or rather, not by that action but somewhere else, since `result.success === false` would ruin the migration process, so it should stop the `init()`?

Reply:
It is possible to have the cancelled state still – during migration :(